File: windows/Helper/PuzzleRepository.cs
using System.Globalization;
using SequenceIQ.Model;

namespace SequenceIQ.Helper;

public static class PuzzleRepository {
    private static readonly Lazy<Dictionary<(Difficulty,int),Puzzle>> Data = new(Load);
    public static Puzzle Get(Difficulty difficulty, int level) => Data.Value.TryGetValue((difficulty,level), out var puzzle) ? puzzle : throw new InvalidOperationException($"Puzzle not found: {difficulty} {level}");
    private static Dictionary<(Difficulty,int),Puzzle> Load() {
        var path = Path.Combine(AppContext.BaseDirectory, "Puzzles.txt");
        var result = new Dictionary<(Difficulty,int),Puzzle>();
        foreach (var line in File.ReadLines(path)) {
            if (string.IsNullOrWhiteSpace(line) || line.StartsWith('#')) continue;
            var p=line.Split('|'); if(p.Length!=5) continue;
            var difficulty=Enum.Parse<Difficulty>(p[1], true);
            var level=int.Parse(p[0][1..], CultureInfo.InvariantCulture);
            var sequence=p[2].Split(',').Select(x => x.Trim()=="?" ? (int?)null : int.Parse(x,CultureInfo.InvariantCulture)).ToArray();
            var options=p[3].Split(',').Select(x=>int.Parse(x,CultureInfo.InvariantCulture)).ToArray();
            var answer=int.Parse(p[4],CultureInfo.InvariantCulture);
            result[(difficulty,level)]=new Puzzle(difficulty,level,sequence,answer,options);
        }
        if(result.Count!=60) throw new InvalidDataException($"Expected 60 puzzles, found {result.Count}.");
        return result;
    }
}

File: android/Helper/PuzzleRepository.kt
package com.hindtechgroup.sequenceiq.Helper

import android.content.Context
import com.hindtechgroup.sequenceiq.Model.Difficulty
import com.hindtechgroup.sequenceiq.Model.Puzzle

object PuzzleRepository {
    fun load(context: Context): List<Puzzle> = context.assets.open("Puzzles.txt").bufferedReader().useLines { lines ->
        lines.mapNotNull { line ->
            if (line.isBlank() || line.startsWith("#")) return@mapNotNull null
            val p=line.split('|'); if(p.size!=5) return@mapNotNull null
            val difficulty=runCatching { Difficulty.valueOf(p[1].uppercase()) }.getOrNull() ?: return@mapNotNull null
            val level=p[0].drop(1).toIntOrNull() ?: return@mapNotNull null
            val sequence=p[2].split(',').map { it.trim().let { v -> if(v=="?") null else v.toIntOrNull() } }
            val options=p[3].split(',').mapNotNull { it.trim().toIntOrNull() }
            val answer=p[4].toIntOrNull() ?: return@mapNotNull null
            if(sequence.size!=6 || options.size!=4) null else Puzzle(difficulty,level,sequence,answer,options)
        }.toList()
    }
}
